internal/docstore/firedocstore/fs.go

```diff
@@ -0,0 +1,516 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package firedocstore provides an implementation of the docstore API for Google
+// Cloud Firestore.
+package firedocstore // import "gocloud.dev/internal/docstore/firedocstore"
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"reflect"
+	"regexp"
+	"strings"
+
+	vkit "cloud.google.com/go/firestore/apiv1"
+	"gocloud.dev/internal/docstore"
+	"gocloud.dev/internal/docstore/driver"
+	"gocloud.dev/internal/gcerr"
+	pb "google.golang.org/genproto/googleapis/firestore/v1"
+	"google.golang.org/grpc/metadata"
+)
+
+type collection struct {
+	client    *vkit.Client
+	dbPath    string
+	collPath  string
+	nameField string
+}
+
+// OpenCollection creates a *docstore.Collection representing a Firestore collection.
+// firedocstore requires that a single string field, nameField, be designated the
+// primary key. Its values must be unique over all documents in the collection, and
+// the primary key must be provided to retrieve a document.
+func OpenCollection(client *vkit.Client, projectID, collPath, nameField string) *docstore.Collection {
+	return docstore.NewCollection(newCollection(client, projectID, collPath, nameField))
+}
+
+func newCollection(client *vkit.Client, projectID, collPath, nameField string) *collection {
+	dbPath := fmt.Sprintf("projects/%s/databases/(default)", projectID)
```

Comment:
Is that always "(default)"? Seems weird.

Reply:
Agreed. I asked the Firestore folks a few times if they wanted the client to support multiple databases per project. They always said no.